core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -312,6 +312,55 @@ public ShardIterator primaryFirstActiveInitializingShardsIt() {
         return new PlainShardIterator(shardId, ordered);
     }
 
+    public ShardIterator replicaActiveInitializingShardIt() {
+        // If the primaries are unassigned, return an empty list (there aren't
+        // any replicas to query anyway)
+        if (!primaryAsList.isEmpty() && !primaryAsList.get(0).active() && !primaryAsList.get(0).initializing()) {
+            List<ShardRouting> emptyList = ImmutableList.of();
+            return new PlainShardIterator(shardId, emptyList);
+        }
+
+        ArrayList<ShardRouting> activeReplicas = new ArrayList<>();
+        ArrayList<ShardRouting> initializingReplicas = new ArrayList<>();
+        ArrayList<ShardRouting> ordered = new ArrayList<>(activeReplicas.size());
```

Comment:
optimization nit, but maybe we can just have one list, and reorder to push active one to the start, we do something similar in primaryFirst. This will mean we don't have to create 3 lists, just one